### PR TITLE
Add allowed_response_headers

### LINKED ISSUE #5031: Sending back a HTTP/1.1 401 Unauthorized with custom header from auth plugin
Hi,

I'm just trying to improve https://github.com/wintoncode/vault-plugin-auth-kerberos and add support for standard SPNEGO auth.

SPNEGO looks something like this:
```
C: POST v1/auth/kerberos/login

S: HTTP/1.1 401 Unauthorized
S: WWW-Authenticate: Negotiate

C: POST v1/auth/kerberos/login
C: Authorization: Negotiate a87421000492aa874209af8bc028
```

Right now we support only the last bit, so if a client includes a valid Authorization header directly in the first request we return a vault token.

This is easy to implement in some languages that let you control the kerberos process enough, but hard in others that expect the full handshake.

The bit I'm missing is a way to send back the 401 with header WWW-Authenticate: Negotiate for the first request. Would it be possible to add support for this in vault?

See also https://github.com/wintoncode/vault-plugin-auth-kerberos/issues/13.

### PR DESCRIPTION
**Pull request by jefferai**
Similar to passthrough_request_headers but going the other way.

Fixes #5031

### CODE REVIEW

File: http/handler.go [hunk 1]
```diff
@@ -577,6 +577,34 @@ func request(core *vault.Core, w http.ResponseWriter, rawReq *http.Request, r *l
 		return nil, false, true
 	}
 
+	if resp != nil && len(resp.Headers) > 0 {
+		// Set this here so it will take effect regardless of any other type of
+		// response processing
+		if len(resp.Headers) > 0 {
```
Comment (kalafut) [inline http/handler.go]: This length check is redundant with the one above.

Comment (jefferai) [reply http/handler.go]: Thanks, fixed!

File: http/handler.go [hunk 2]
```diff
@@ -664,26 +664,28 @@ func respondStandby(core *vault.Core, w http.ResponseWriter, reqURL *url.URL) {
 
 // getTokenFromReq parse headers of the incoming request to extract token if present
 // it accepts Authorization Bearer (RFC6750) and X-Vault-Token header
-func getTokenFromReq(r *http.Request) (string, error) {
+func getTokenFromReq(r *http.Request) (string, bool, error) {
```
Comment (kalafut) [inline http/handler.go]: This `error` response is always nil and could be removed.

Comment (jefferai) [reply http/handler.go]: Done.

File: http/handler.go [hunk 3]
```diff
@@ -664,26 +664,28 @@ func respondStandby(core *vault.Core, w http.ResponseWriter, reqURL *url.URL) {
 
 // getTokenFromReq parse headers of the incoming request to extract token if present
 // it accepts Authorization Bearer (RFC6750) and X-Vault-Token header
```
[the remark this answers is not in the document]
Comment (jefferai) [reply http/handler.go]: Done.

File: http/handler_test.go [hunk 4]
```diff
@@ -601,36 +601,41 @@ func TestHandler_requestAuth(t *testing.T) {
...
 	} else if tok != "" {
 		t.Fatalf("expected '' as result, got '%s'", tok)
 	}
 
 	r.Header.Set("Authorization", "Bearer TOKEN NOT_GOOD_TOKEN")
-	if tok, err := getTokenFromReq(&r); err == nil {
-		t.Fatalf("expected an error, got none")
-	} else if tok != "" {
-		t.Fatalf("expected '' as result, got '%s'", tok)
+	if _, fromHeader, err := getTokenFromReq(&r); err != nil {
```
[the remark this answers is not in the document]
Comment (jefferai) [reply http/handler_test.go]: Done.

**Pull request merged**